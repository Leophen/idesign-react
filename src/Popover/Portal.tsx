import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import ReactDOM from 'react-dom';
import { placementType, PortalProps } from './type';
import { useContainer } from '../hooks/useContainer';

// 创建气泡提示容器
const popoverWrapper = useContainer('i-popover-wrapper', document.body);

const Portal: React.FC<PortalProps> = (props) => {
  const {
    className,
    style,
    visible = false,
    content = '',
    placement = 'top',
    getRef,
    ...tProps
  } = props;

  const getLocationStyle = (
    placement: placementType,
    trigger: PortalProps,
    popover: PortalProps,
  ) => {
    let popoverWidth = style?.width ? Number(style?.width) : popover.width;
    let popoverHeight = style?.height ? Number(style?.height) : popover.height;
    const xMap = {
      top: trigger.left + (trigger.width - popoverWidth) / 2,
      'top-left': trigger.left,
      'top-right': trigger.left + (trigger.width - popoverWidth),
      bottom: trigger.left + (trigger.width - popoverWidth) / 2,
      'bottom-left': trigger.left,
      'bottom-right': trigger.left + (trigger.width - popoverWidth),
      left: trigger.left - popoverWidth - 16,
      'left-top': trigger.left - popoverWidth - 16,
      'left-bottom': trigger.left - popoverWidth - 16,
      right: trigger.left + trigger.width + 16,
      'right-top': trigger.left + trigger.width + 16,
      'right-bottom': trigger.left + trigger.width + 16,
    };
    const yMap = {
      top: trigger.top - popoverHeight - 16,
      'top-left': trigger.top - popoverHeight - 16,
      'top-right': trigger.top - popoverHeight - 16,
      bottom: trigger.top + trigger.height + 16,
      'bottom-left': trigger.top + trigger.height + 16,
      'bottom-right': trigger.top + trigger.height + 16,
      left: trigger.top + (trigger.height - popoverHeight) / 2,
      'left-top': trigger.top,
      'left-bottom': trigger.top + (trigger.height - popoverHeight),
      right: trigger.top + (trigger.height - popoverHeight) / 2,
      'right-top': trigger.top,
      'right-bottom': trigger.top + (trigger.height - popoverHeight),
    };
    const result = {
      left: xMap[placement],
      top: yMap[placement],
      ...style,
    };
    return result;
  };

  const popoverRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    getRef?.(popoverRef);
  }, []);

  // 重置方向次数限制
  const rePlaceNum = useRef(0);

  // 更新气泡方向
  const updatePlacement = (currentPlacement: placementType) => {
    if (popoverRef.current) {
      // 原触发方向
      const direction = currentPlacement.split('-')[0];
      const directionWith = currentPlacement.split('-')[1]
        ? '-' + currentPlacement.split('-')[1]
        : '';
      // 窗口
      const winWidth = window.innerWidth;
      const winHeight = window.innerHeight;
      // 气泡
      const rect = popoverRef.current.getBoundingClientRect();
      const popoverWidth = rect.width;
      const popoverHeight = rect.height;
      const popoverTop = rect.top;
      const popoverLeft = rect.left;

      let result: placementType = currentPlacement;
      if (popoverTop < winHeight && popoverLeft < winWidth && rePlaceNum.current < 3) {
        if (direction === 'top' && popoverTop < 0) {
          result = ('bottom' + directionWith) as placementType;
        }
        if (direction === 'bottom' && winHeight - popoverHeight - popoverTop < 0) {
          result = ('top' + directionWith) as placementType;
        }
        if (direction === 'left' && popoverLeft < 0) {
          result = ('right' + directionWith) as placementType;
        }
        if (direction === 'right' && winWidth - popoverWidth - popoverLeft < 0) {
          result = ('left' + directionWith) as placementType;
        }
        rePlaceNum.current += 1;
      }

      setCurrentPlacement(result);
    }
  };

  const [styles, setStyles] = useState({});
  const [currentPlacement, setCurrentPlacement] = useState<placementType>(placement);

  useEffect(() => {
    updatePlacement(currentPlacement);
  });

  useEffect(() => {
    const rect = (popoverRef.current as HTMLElement).getBoundingClientRect();
    setStyles(getLocationStyle(currentPlacement, { ...tProps }, rect));
  }, [visible, currentPlacement, tProps.width, tProps.height, tProps.left, tProps.top]);

  const PopoverNode = (
    <div
      ref={popoverRef}
      className={classNames('i-popover', className)}
      data-popper-placement={currentPlacement}
      style={{ ...styles, width: style?.width, height: style?.height }}
    >
      <div className="i-popover__arrow" data-popper-placement={currentPlacement} />
      {typeof content === 'object' ? content : <div className="i-popover__text">{content}</div>}
    </div>
  );

  return ReactDOM.createPortal(PopoverNode, popoverWrapper as HTMLElement);
};

export default Portal;
