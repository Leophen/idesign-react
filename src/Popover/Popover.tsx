import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Transition from '../Transition';
import useDefault from '../hooks/useDefault';
import Portal from './Portal';
import { PopoverProps } from './type';
import { useHasParent } from '../hooks/useHasParent';

// 原 Popup 原生实现写法
const Popover: React.FC<PopoverProps> = (props) => {
  const {
    children = '',
    className,
    portalClassName = '',
    style,
    content,
    placement = 'top',
    trigger = 'hover',
    visible,
    defaultVisible = false,
    disabled = false,
    updateLocation,
    onTrigger = () => { }
  } = props;

  const [top, setTop] = useState(0)
  const [left, setLeft] = useState(0)
  const [width, setWidth] = useState(0)
  const [height, setHeight] = useState(0)

  const setTargetLocation = (target: HTMLElement) => {
    const rect = target.getBoundingClientRect()
    setTop((rect?.top || 0) + window.scrollY)
    setLeft((rect?.left || 0) + window.scrollX)
    setWidth((rect?.width || 0))
    setHeight((rect?.height || 0))
  }

  const [innerVisible, setInnerVisible] = useDefault(visible, defaultVisible, onTrigger);

  // 手动更新实时位置
  useEffect(() => {
    const currentTriggerNode = (triggerNode.current as any).children[0]
    setTargetLocation(currentTriggerNode)
  }, [updateLocation])

  // 打开气泡通用方法
  const switchPopover = (e: React.MouseEvent, show: boolean) => {
    // 设置气泡位置
    // let currentTriggerNode: any = e.target
    // while (currentTriggerNode?.parentNode === (triggerNode.current as unknown as HTMLElement).children[0]) {
    //   currentTriggerNode = currentTriggerNode?.parentNode
    // }
    // if (currentTriggerNode === triggerNode.current) {
    //   currentTriggerNode = currentTriggerNode.children[0]
    // }
    // console.log(currentTriggerNode.parentNode,triggerNode.current.children[0])
    const currentTriggerNode = (triggerNode.current as any).children[0]
    if (show) {
      e.persist();
      setTargetLocation((currentTriggerNode))
    }
    // 设置气泡显示隐藏
    setInnerVisible(show)
  }

  // 关闭气泡通用方法
  const closePopover = () => {
    setInnerVisible(false)
  }

  // 气泡包裹层节点
  const triggerNode = useRef<HTMLDivElement>(null)

  // 全局监听事件，判断点击节点是否在气泡内，以确定是否关闭气泡
  const [listenClick, setListenClick] = useState(false)
  const ifClickInPopover = (e: any) => {
    // 点击位置在气泡外
    if (!useHasParent(e.target, popoverRef.current)) {
      // 点击位置既在气泡外 又在触发节点外
      if (!useHasParent(e.target, triggerNode?.current as HTMLElement)) {
        closePopover()
      }
      setListenClick(false)
      window.removeEventListener('click', ifClickInPopover)
    }
  }
  useEffect(() => {
    if (listenClick) {
      window.addEventListener('click', ifClickInPopover)
      return () => window.removeEventListener('click', ifClickInPopover)
    }
  }, [listenClick])
  // 点击触发节点
  const handleClick = (e: React.MouseEvent) => {
    if (trigger === 'click') {
      switchPopover(e, !innerVisible)
      // 气泡在关闭状态下点击 则监听下一次全局点击事件
      if (!innerVisible) {
        setTimeout(() => setListenClick(true))
      }
    } return
  }

  const popoverRef = useRef<any>(null)
  const getPopoverRef = (ref: any) => {
    popoverRef.current = ref.current as HTMLElement
  }

  // 判断点击和右击节点是否在气泡内
  const [listenContextMenu, setListenContextMenu] = useState(false)
  const ifHandleInPopover = (e: any) => {
    e.preventDefault();
    if (!useHasParent(e.target, popoverRef.current as any)) {
      closePopover()
      setListenContextMenu(false)
      window.removeEventListener('click', ifHandleInPopover)
      window.removeEventListener('contextmenu', ifHandleInPopover)
    }
  }
  useEffect(() => {
    if (listenContextMenu) {
      window.addEventListener('click', ifHandleInPopover)
      window.addEventListener('contextmenu', ifHandleInPopover)
      return () => {
        window.removeEventListener('click', ifHandleInPopover)
        window.removeEventListener('contextmenu', ifHandleInPopover)
      }
    }
  }, [listenContextMenu])
  // 右击触发节点
  const handleContextMenu = (e: React.MouseEvent) => {
    if (trigger === 'context-menu') {
      e.preventDefault();
      switchPopover(e, !innerVisible)
      if (!innerVisible) {
        setTimeout(() => setListenContextMenu(true))
      }
    } return
  }

  // 判断悬浮节点是否在气泡内
  const ifHoverInPopover = (e: any) => {
    e.preventDefault();
    // 悬浮位置在气泡外
    if (!useHasParent(e.target, popoverRef.current)) {
      // 悬浮位置既在气泡外 又在触发节点外
      if (!useHasParent(e.target, triggerNode?.current as HTMLElement)) {
        closePopover()
      }
      window.removeEventListener('click', ifClickInPopover)
    }
  }

  // 悬浮触发节点
  const handleEnter = (e: React.MouseEvent) => {
    if (trigger === 'hover') {
      switchPopover(e, !innerVisible)
      setTimeout(() => {
        window.addEventListener('mouseover', ifHoverInPopover)
      })
    } return
  }

  // 触发节点宽高变化时重定位
  useEffect(() => {
    if (typeof ResizeObserver !== 'undefined') {
      const resizeObserver = new ResizeObserver(entries => {
        setWidth((entries[0].contentRect.width || 0))
        setHeight((entries[0].contentRect.height || 0))
      });
      resizeObserver.observe((triggerNode.current as any))
      return () => resizeObserver.disconnect()
    }
  }, [])

  return (
    <div
      className={classNames(
        'i-popover__reference',
        className
      )}
    >
      <section
        onClick={handleClick}
        onMouseEnter={handleEnter}
        onContextMenu={handleContextMenu}
        ref={triggerNode}
      >
        {children}
      </section>

      <Transition
        timeout={200}
        in={innerVisible && !disabled}
        animation='fade-in'
        key='i-popover'
      >
        <Portal
          style={{ ...style }}
          className={portalClassName}
          visible={innerVisible && !disabled}
          content={content}
          placement={placement}
          top={top}
          left={left}
          width={width}
          height={height}
          getRef={getPopoverRef}
        />
      </Transition>
    </div>
  );
};

Popover.displayName = 'Popover';

export default Popover;
