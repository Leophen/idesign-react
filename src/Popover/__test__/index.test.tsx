import { render } from '@testing-library/react';
import Popover from '../Popover';

describe('Popover 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Popover content="提示内容">
        <button>悬浮提示</button>
      </Popover>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

