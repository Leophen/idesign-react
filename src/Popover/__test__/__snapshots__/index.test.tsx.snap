// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-popover__reference"
  >
    <section>
      <button>
        悬浮提示
      </button>
    </section>
  </div>
</DocumentFragment>
`;
